middle_end/flambda2/docs/backend_exceptions.md

```diff
@@ -0,0 +1,241 @@
+
+Exceptions in the OCaml language are manipulated through two constructions:
+the `raise` primitive, and the `try...with` structure (`match...with exception`
+is considered here as syntactic sugar for and `try...with` around a regular
+`match`).
+
+In the native backend, this is implemented using trap frames on the stack, and
+a global variable (now part of the `Domain_state` structure) to hold a pointer
+to the current (or top) frame.
+A trap frame is a small structure holding a pointer to the previous trap frame
+and a code pointer, and it is stored on the regular stack, along with spilled
+local variables..
+
+The compilation of a `raise` primitive consists of storing
+the exception being raised into the appropriate register, then loading the
+current trap frame through the global variable, restoring the value of the current
+trap frame to the previous value, restoring the value of the stack pointer to
+just before (or just after, for backwards-growing stacks) the trap frame, and
+finally jumping to the code pointer contained in the stack frame.
+
+The compilation of a `try...with` structure will produce code that:
+- Prefixes the handler's code with a label that can be used for computing its code
+pointer.
+- Prefixes the body's code with a set of instructions to allocate the trap frame.
+This is subsumed in the `Linear` representation by the `Lpushtrap` instruction.
+- Appends to the body a set of instructions to deallocate the trap frame.
+This is subsumed in the `Linear` representation by the `Lpoptrap` instruction.
+
+So before the `try...with` there is a given stack shape that we will call the
+base stack, the body is executed with a stack shape that is the base stack plus
+a trap frame, after the body (if the body doesn't raise) the execution resumes
+with the base stack, and if the body raises execution goes to the handler, with
+the base stack (and when the handler finishes execution flows to the same code
+as after the body, with the base stack).
+
+## Problematic cases
+
+This all works well as long as all code paths go through the `Lpushtrap` and
+`Lpoptrap` instructions in a well-parenthesized manner. However, this invariant
+can be broken by the static exceptions used in the compiler representations
+from Lambda and onwards. In particular, although it only occurs in specific
+circumstences, it is allowed to have inside the body of a `try...with`
+structure a sub-expression that raises a static exception whose handler is
+outside the body of the `try...with`. This looks like a simple detail, as
+the code that emits the traps simply needs to insert traps around these
+static raising constructs, but this is actually also important for earlier
+passes as some optimisations involving static exceptions that would be correct
+if both the handler and the raise are in the same exception context become
+incorrect if the contexts differ.
+
+To deal with this, the relevant optimisations in the compiler carry a `try_depth`
```

Comment:
(Not sure it is worth mentioning, but another consequence is that
we need to generate multiple `Lpoptrap` in such a situation.)

Reply:
I've edited the doc to add a mention of multiple traps.